### ISSUE TITLE
Android Version

### ISSUE THREAD

**Issue by sungshon** [sungshon/PimpMyStremio#51]
I've done some work on this and wanted to track my progress and current blocking issues regarding it if others may wish to join on the discussion.

I've managed to get PimpMyStremio working on Android 64bit for now, currently lacking `youtube-dl` and `phantomjs` support.

I've accomplished this by using:
https://github.com/node-on-mobile/node-on-android

One major issue at this point is that `node-on-android` only has a `node.js` build for 64bit, no 32bit yet. I've found some sources for a 32bit build of node.js for android so I'm hopeful, otherwise I'm also looking at this project as an alternative:
https://code.janeasystems.com/nodejs-mobile

The second major issue is `youtube-dl` and `phantomjs`.
I do believe `youtube-dl` can be used on android with termux packages:
https://github.com/termux/termux-packages

And then a bridge to it can be exposed to node.js.

But `phantomjs` is not available in termux packages:
https://github.com/termux/termux-packages/issues/908

I've seen cases where devs successfully use WebView on android as a headless browser, but that would require devs to write separate code for desktop / android to make their add-ons support android too.

With this said, I'm considering making a `supportsAndroid` setting in the add-on list, and setting all add-ons that use `phantomjs` as not android ready.

This is my current progress with it, but I dare say currently all add-ons that don't need binaries work beautifully atm.

**Comment by chicungunya** [comment]
Hello,
Jackett works?
Thanks a lot

**Comment by sungshon** [comment]
@chicungunya There is no official android version yet, this ticket is for the work being done on a future official android version of PimpMyStremio.

**Comment by sungshon** [comment]
Further research shows:
- 5 / 43 addons are using `phantomjs`
- 9 / 43 addons are using `youtube-dl`
- 2 addons are using both `phantomjs` and `youtube-dl`

**Comment by sungshon** [comment]
Also to note I'm looking into [Zombie.js](https://www.npmjs.com/package/zombie) as a possible replacement for `phantomjs`, but it's based on `jsdom`, so I'm not very hopeful for now..

**Comment by imurreflection** [comment]
Hi Sungshon..thanks a lot for working on making PMS available on Android. This is a much awaited add-on.

**Comment by sungshon** [comment]
@imurreflection It will take some time, but it's possible.

**Comment by sungshon** [comment]
Notes:
- added [Zombie.js](https://www.npmjs.com/package/zombie) to module whitelist as a possible alternative to phantomjs
- there is also a `zombie.js` shim for `phantomjs` here: https://www.npmjs.com/package/zombie-phantom (a shim is a really really bad solution, just noting it for future reference)

**Comment by sungshon** [comment]
Seems [Zombie.js](https://www.npmjs.com/package/zombie) can't be used as a replacement for `phantomjs`, this doesn't surprise me.

I'm fairly convinced by now that "fake browsers" simply can't be used to replace headless browsers.

Current plan:
- completely replace `phantomjs` with `puppeteer` (see [this ticket](https://github.com/sungshon/PimpMyStremio/issues/41) for further information)
- for the android version we will use `youtube-dl` and `puppeteer` remotely (they will be deployed to AWS Lambda)

Issues with this plan:
- it will break the launcher (updater), because it checks if `phantomjs` was correctly downloaded, so all users will need to redownload the launcher...
- the API of `puppeteer` would need to be reconstructed so it can use the remote version (might not be fully possible, as it uses promises and we can only make 1 http request)
- on android, due to using remote versions of `youtube-dl` and `puppeteer`, all IP bound streams will break, so it will definitely break _some_ addons

Benefits:
- although it would break some addons, i do expect some addons to still work, so we're not loosing all 12 addons (that use these binaries)
- even if the `puppeteer` API can't be completely reproduced, it might be enough anyway
- `puppeteer` is almost 300mb in size, which is HUGE, but if I webpack all the node modules of the binary dependencies, we could save some serious space from the other binary requirements, in the end the size difference of the update might not be drastically different
- this might work for iOS too, not just android

I think this is the most plausible plan for this.. because it removes all binary dependencies completely, and if they don't exist, the application will just try to use the remote versions as alternatives. Simplifying PimpMyStremio to this level might bring even more benefits in the future.

**Comment by sungshon** [comment]
Rebuilding the `puppeteer` API to seem like it's working locally but actually work remotely is not a solution. There are many reasons this can't be done correctly and I don't see a reason to go into them.

So the current plan is now:
- use `youtube-dl` remotely for Android / iOS
- block using the `phantomjs` locally completely on Android / iOS
- suggest deploying (to now.sh) `puppeteer` and / or `phantomjs` logic to addon developers, and using the remote deployments instead of local ones if possible
- create docs about how to use `puppeteer` and / or `phantomjs` on now.sh deployments

I'm satisfied with this solution, it seems to be the only way for now.

**Comment by nsidhdi** [comment]
Thank for this so what is the progress now?

**Comment by sleeyax** [comment]
Here's my attempt at this: [sleeyax/android-pms](https://github.com/sleeyax/android-pms).

It uses [androidjs](https://github.com/android-js/androidjs), which is basically a webview bundled with the Node runtime.

**Comment by Baricanac** [comment]
After becoming acquainted with Stremio on my Google Chromecast with Google TV and after that finding out about PMS and Jackett. I've started to wonder why the community add-ons handled by PMS or Jacket weren't just made part of the Stremio community add-ons like most of the regulars on there. Any particular reason why you have had to end up developing PMS as a workaround to cater to these needs? I've got to say that it would be incredible if PMS and Jacket could be installed as just another add-on on top of Stremio even for the Android version in the long run.

**Comment by sleeyax** [comment]
@Baricanac 

Stremio doesn't support local addons (like kodi for example) so PMS had to be developed as a work around. All Stremio addons you can install from the catalog actually run on a remote HTTP webserver.   

Some addons have specific needs that won't work when served from a single remote web server (e.g IP address rate limits, custom headers that need to be sent or rotated).  

In theory the jacket addon would work remotely (I've tried this!) but it just takes a long time to load sources, which would put a lot of stress on the server. You'll also need to keep jackett up to date so in the long run it's better to have it as a local addon.

**Comment by Baricanac** [comment]
@sleeyax Thanks for clearing that up. 

I ended up finding out about your unofficial PMS app for Android (version 0.0.3-beta) and have managed to install it on my Chromecast 2020. So as for now I am currenlty testing it out. Although it is still a bit unclear to me on how the workflow is supposed to operate due to the limited documentation provided. 

I did manage to install some of the add-ons provided by your (unofficial) Android PMS app on to the official Stremio Android app (version 1.3.4) which weren't part of the official/community Stremio repositories. Although I haven't confirmed any of them to be working yet. For an example, I found out that the Drama Very add-on seems obselete by now due to the verydrama.com website being offline? I'll keep you posted on the official github page for the unofficial PMS-Android app if I find out anything new.

**Comment by sleeyax** [comment]
@Baricanac 

Yes, some (if not most) local PMS addons are outdated. You can open an issue on their respective GitHub repositories and ask for an update (or removal if the site is dead) by the author.
